Adds (Untested) Add Delete Load Endpoints

hipsterusername opened the pull request:
This is probably not working code, but it is an attempt.

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+    operation_id="add_model",
+    responses={201: {"model": Union[CkptModelInfo, DiffusersModelInfo], "new_model_list": ModelsList}},
+)
+async def add_model(
```
`set_model` might be a better name, since this is really an "add or replace" operation.

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+""" Add Model """
+@models_router.post(
+    "/",
+    operation_id="add_model",
```
also `set_model`

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+        if len(model_attributes.get("vae", [])) == 0:
+            del model_attributes["vae"]
+
+        ApiDependencies.invoker.services.model_manager.add_model(
```
Any idea how long this takes in the worst case? I'd worry about this method timing out. May make sense to return a `202` if we can detect that it'll need to download a model or convert a model or something.

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+        del model_info["model_name"]
+        model_attributes = model_info
+
+        if len(model_attributes.get("vae", [])) == 0:
```
This is invalid. Pydantic should handle parsing stuff and defaults anyway.

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+    """Adds a new model"""
+    try:
+        model_name = model_info["model_name"]
+        del model_info["model_name"]
```
This is invalid. You shouldn't add `model_name` to `ModelInfo` just to be able to pass it in. Instead, define a new request class that has the model name and model info. Something like:

```py
class CreateModelRequest(BaseModel):
    name: str = Field(description="The name of the model")
    info: Annotated[Union[(CkptModelInfo,DiffusersModelInfo)], Field(discriminator="format")] = Field(description="The model info")
```

(Note: I think you combine the two `Field`s above... add the description inside the annotation)

hipsterusername replied:
If the Model Info doesn't have the Name, how is the Model List info going to have references for it? Is the Name something passed around as a label for the entirety of the `info` object even though it's not explicitly defined?

Additionally, what ID is used to select in order to update the model if not name?

hipsterusername replied:
So - Walking through my thinking to figure out where assumptions are off.

Name seemed to be needed as part of the "Model" object generally. (I didn't see where else it would be defined in order to reference it)

The old/legacy code did the same function as above - Explicitly, for this function, splitting the `info` object into a name and attributes, as the backend code (in model_manager) is specifically looking for the request to be formatted like (model_name, model_attributes, clobber) as far as I understand it.

Recognizing that I could create another class for the request, but also noting that the Name is needed elsewhere (presumably it should be included in the models list?) I figured adding it to the core model info definition would be ideal.

Ok - now I can be educated. :)

Kyle0654 replied:
My read is that the models are stored in a dictionary (or dict-like) structure, where the name is the key and the info is the value. The big issue is this:
```py
        model_name = model_info["model_name"]
        del model_info["model_name"]
```
That would delete the `model_name` property from the `model_info` object, which would make it invalid (since the `model_name` is a required property the way you've added it), and pydantic would throw a fit over it eventually.

hipsterusername replied:
Ahhh. I see. Ok - will need to update this tonight!

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+    responses={201: {"model": Union[CkptModelInfo, DiffusersModelInfo], "new_model_list": ModelsList}},
+)
+async def add_model(
+    model_info: Union[CkptModelInfo, DiffusersModelInfo],
```
You need to use a discriminated union to tell which type the user passed:
```py
Annotated[Union[(CkptModelInfo,DiffusersModelInfo)], Field(discriminator="format")]
```
Though see below - I think you need a wrapping object here.

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+# Kent wrote the below code. It is highly suspect, and should be reviewed. Once all issues have been identified and fixed, this comment can be removed.
+# Seriously, check this. I have no idea how to test it.
+
+""" Add Model """
```
You don't need triple-quote comments above things =)

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -54,6 +55,90 @@ async def list_models() -> ModelsList:
+        raise HTTPException(status_code=500, detail=str(e))
+
+
+""" Load Model """
```
Remove this method - we don't explicitly load a model anymore - it's all done automatically through nodes now.

hipsterusername commented:
@Kyle0654 - I made some updates based on your review. There's likely _still_ some cleanup to do, but I'm pleased it wasn't just completely a "burn it down" situation. 🎉

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -55,30 +55,28 @@ async def list_models() -> ModelsList:
+    responses={
+        201: {
+        "model": Union[CkptModelInfo, DiffusersModelInfo], 
+        "new_model_list": ModelsList
```
I'd hesitate to return the entire list here.

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -55,30 +55,28 @@ async def list_models() -> ModelsList:
-    """Adds a new model"""
+async def update_model(
+    model_request: CreateModelRequest
+) -> CreateModelRequest:
```
You probably need a `CreateModelResponse` here (after the `->` is the return type).

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -55,30 +55,28 @@ async def list_models() -> ModelsList:
+async def update_model(
+    model_request: CreateModelRequest
+) -> CreateModelRequest:
+    #Adds a new model
```
This should be triple-quoted to show up in the API docs (launch and go to `http://localhost:9090/docs` to see what the API docs will look like for your endpoints).

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -55,30 +55,28 @@ async def list_models() -> ModelsList:
         ApiDependencies.invoker.services.model_manager.add_model(
-            model_name=model_name,
-            model_attributes=model_attributes,
+            model_name=model_request["name"],
```
You should use `model_request.name` and `model_request.info`, since pydantic has verified everything.

hipsterusername commented:
@Kyle0654 - I'll be honest, I didn't realize what the triple quotes were doing, now I do! :) 

Also, appreciate the pointers on things. Picking up some things but... still largely doing monkey see monkey do.

Updates in!

Kyle0654 commented on invokeai/app/api/routers/models.py:
```diff
@@ -71,22 +75,23 @@ async def list_models() -> ModelsList:
+            model_attributes=model_request.info,
             clobber=True,
         )
-        # How does Ckpt support deprecation change the above?
```
Your `CreateModelResponse` class has `name` and `info` fields, no `status` field. You'll need to provide those, or change the response structure. I'd also recommend you try to run each of these endpoints through the openapi doc (swagger ui) to see if they work =). The code looks okay, but I'm not sure if it'll work.

Kyle0654 reviewed (approved):
Approving, but with a comment that needs to be resolved before merge. Also, I recommend trying to hit each of these endpoints to see if they work.

Kyle0654 commented:
> still largely doing monkey see monkey do.

I'm of the opinion that code code should enable contribution in that way, so that's a compliment =)

lstein commented:
This looks like Kyle has approved for merge but there are many unresolved conversations. Should this be merged?

Kyle0654 commented:
Ah do we not have "comments must be resolved" as a requirement for merging? If I approve but leave comments, that usually indicates "these are things that need to be addressed, and then I approve". Especially given time zone differences, it cuts down on back and forth time.

hipsterusername commented:
> Ah do we not have "comments must be resolved" as a requirement for merging? If I approve but leave comments, that usually indicates "these are things that need to be addressed, and then I approve". Especially given time zone differences, it cuts down on back and forth time.

I guess the question is - Have my changes addressed the comments or not? I believe I have pushed updates for all of them, to the best of my abilities. I can try to do a convert and merge endpoint once these two are 👍

Kyle0654 commented:
On my phone for a while, but if you have a modelInfo class in there still, you'd at least need to change that to ModelInfo. Otherwise, if you run the server and try the endpoints and they work, it should be good to go.

hipsterusername commented:
> On my phone for a while, but if you have a modelInfo class in there still, you'd at least need to change that to ModelInfo. Otherwise, if you run the server and try the endpoints and they work, it should be good to go.

My challenge w/ testing is not being able to get docs to open.

  File "X:\stablediffusion\InvokeAI-files\.venv\lib\site-packages\fastapi\utils.py", line 45, in get_model_definitions
    m_schema, m_definitions, m_nested_models = model_process_schema(
  File "pydantic\schema.py", line 581, in pydantic.schema.model_process_schema
  File "pydantic\schema.py", line 622, in pydantic.schema.model_type_schema
  File "pydantic\schema.py", line 255, in pydantic.schema.field_schema
  File "pydantic\schema.py", line 527, in pydantic.schema.field_type_schema
  File "pydantic\schema.py", line 927, in pydantic.schema.field_singleton_schema
  File "pydantic\schema.py", line 581, in pydantic.schema.model_process_schema
  File "pydantic\schema.py", line 622, in pydantic.schema.model_type_schema
  File "pydantic\schema.py", line 255, in pydantic.schema.field_schema
  File "pydantic\schema.py", line 480, in pydantic.schema.field_type_schema
  File "pydantic\schema.py", line 850, in pydantic.schema.field_singleton_schema
  File "pydantic\schema.py", line 701, in pydantic.schema.field_singleton_sub_fields_schema
  File "pydantic\schema.py", line 527, in pydantic.schema.field_type_schema
  File "pydantic\schema.py", line 850, in pydantic.schema.field_singleton_schema
  File "pydantic\schema.py", line 745, in pydantic.schema.field_singleton_sub_fields_schema
  File "pydantic\schema.py", line 527, in pydantic.schema.field_type_schema
  File "pydantic\schema.py", line 925, in pydantic.schema.field_singleton_schema
KeyError: <class 'pydantic.main.LoadImageInvocation'>

I am getting other pydantic errors even when commenting out what I added (in the same file.) 

Without the docs, not exactly sure how to test the endpoints (Getting 404 Not Found's on all the endpoint combos I'm trying on the GET for even the models list) - Thanks for the patience in dealing with my noobish issues. :)

Kyle0654 commented:
Try rebasing on main. There was a recent update to the FastAPI version, since it was causing issues for some users.

hipsterusername commented:
Getting weird Rebasing issues. Says its rebased though - Although, getting socketio errors of "404 Not Found" now... 🙈
